core/src/main/java/jenkins/model/Nodes.java

```diff
@@ -0,0 +1,282 @@
+/*
+ * The MIT License
+ *
+ * Copyright (c) 2015, CloudBees, Inc., Stephen Connolly
+ *
+ * Permission is hereby granted, free of charge, to any person obtaining a copy
+ * of this software and associated documentation files (the "Software"), to deal
+ * in the Software without restriction, including without limitation the rights
+ * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+ * copies of the Software, and to permit persons to whom the Software is
+ * furnished to do so, subject to the following conditions:
+ *
+ * The above copyright notice and this permission notice shall be included in
+ * all copies or substantial portions of the Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+ * THE SOFTWARE.
+ */
+package jenkins.model;
+
+import hudson.BulkChange;
+import hudson.Util;
+import hudson.XmlFile;
+import hudson.model.Computer;
+import hudson.model.ItemGroupMixIn;
+import hudson.model.Node;
+import hudson.model.Queue;
+import hudson.model.Saveable;
+import hudson.model.listeners.SaveableListener;
+import hudson.slaves.EphemeralNode;
+import hudson.slaves.OfflineCause;
+import org.kohsuke.accmod.Restricted;
+import org.kohsuke.accmod.restrictions.NoExternalUse;
+
+import javax.annotation.CheckForNull;
+import javax.annotation.Nonnull;
+import java.io.File;
+import java.io.FileFilter;
+import java.io.IOException;
+import java.util.ArrayList;
+import java.util.Collection;
+import java.util.HashSet;
+import java.util.Iterator;
+import java.util.List;
+import java.util.Map;
+import java.util.Set;
+import java.util.TreeMap;
+import java.util.concurrent.ConcurrentMap;
+import java.util.concurrent.ConcurrentSkipListMap;
+import java.util.logging.Level;
+import java.util.logging.Logger;
+
+/**
+ * Manages all the nodes for Jenkins.
+ *
+ * @since 1.FIXME
+ */
+@Restricted(NoExternalUse.class) // for now, we may make it public later
+public class Nodes implements Saveable {
```

Comment:
Should have at least a warning in the changelog that you lose configuration when downgrading Jenkins. [JENKINS-27700](https://issues.jenkins-ci.org/browse/JENKINS-27700) rightfully complains.